libbpf-tools/filetop.c

```diff
@@ -0,0 +1,319 @@
+			sort_by = WRITES;
+		} else if (!strcmp(arg, "rbytes")) {
+			sort_by = RBYTES;
+		} else if (!strcmp(arg, "wbytes")) {
+			sort_by = WBYTES;
+		} else {
+			warn("invalid sort method: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	case 'r':
+		errno = 0;
+		rows = strtol(arg, NULL, 10);
+		if (errno || rows <= 0) {
+			warn("invalid rows: %s\n", arg);
+			argp_usage(state);
+		}
+		output_rows = rows;
+		if (output_rows > OUTPUT_ROWS_LIMIT)
+			output_rows = OUTPUT_ROWS_LIMIT;
+		break;
+	case 'h':
+		argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
+		break;
+	case ARGP_KEY_ARG:
+		errno = 0;
+		if (pos_args == 0) {
+			interval = strtol(arg, NULL, 10);
+			if (errno || interval <= 0) {
+				warn("invalid interval\n");
+				argp_usage(state);
+			}
+		} else if (pos_args == 1) {
+			count = strtol(arg, NULL, 10);
+			if (errno || count <= 0) {
+				warn("invalid count\n");
+				argp_usage(state);
+			}
+		} else {
+			warn("unrecognized positional argument: %s\n", arg);
+			argp_usage(state);
+		}
+		pos_args++;
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static void sig_int(int signo)
+{
+	exiting = 1;
+}
+
+static int sort_column(const void *obj1, const void *obj2)
+{
+	struct file_stat *s1 = (struct file_stat *)obj1;
+	struct file_stat *s2 = (struct file_stat *)obj2;
+
+	if (sort_by == READS) {
+		return s1->reads - s2->reads;
+	} else if (sort_by == WRITES) {
+		return s1->writes - s2->writes;
+	} else if (sort_by == RBYTES) {
+		return s1->read_bytes - s2->read_bytes;
+	} else if (sort_by == WBYTES) {
+		return s1->write_bytes - s2->write_bytes;
+	} else {
+		return (s1->reads + s1->writes + s1->read_bytes + s1->write_bytes)
+		     - (s2->reads + s2->writes + s2->read_bytes + s2->write_bytes);
+	}
+}
+
+static int print_stat(struct filetop_bpf *obj)
+{
+	FILE *f;
+	time_t t;
+	struct tm *tm;
+	char ts[16], buf[256];
+	struct file_id key, *prev_key = NULL;
+	static struct file_stat values[OUTPUT_ROWS_LIMIT];
+	int n, i, err = 0, rows = 0;
+	int fd = bpf_map__fd(obj->maps.entries);
+
+	f = fopen("/proc/loadavg", "r");
+	if (f) {
+		time(&t);
+		tm = localtime(&t);
+		strftime(ts, sizeof(ts), "%H:%M:%S", tm);
+		memset(buf, 0 , sizeof(buf));
+		n = fread(buf, 1, sizeof(buf), f);
+		if (n)
+			printf("%8s loadavg: %s\n", ts, buf);
+		fclose(f);
+	}
+
+	printf("%-7s %-16s %-6s %-6s %-7s %-7s %1s %s\n",
+	       "TID", "COMM", "READS", "WRITES", "R_Kb", "W_Kb", "T", "FILE");
+
+	while (1) {
+		err = bpf_map_get_next_key(fd, prev_key, &key);
+		if (err) {
+			if (errno == ENOENT) {
+				err = 0;
+				break;
+			}
+			warn("bpf_map_get_next_key failed: %s\n", strerror(errno));
+			return err;
+		}
+		err = bpf_map_lookup_elem(fd, &key, &values[rows++]);
+		if (err) {
+			warn("bpf_map_lookup_elem failed: %s\n", strerror(errno));
+			return err;
+		}
+		prev_key = &key;
+	}
+
+	qsort(values, rows, sizeof(struct file_stat), sort_column);
+	rows = rows < output_rows ? rows : output_rows;
```

Comment:
I see what you mean by "the descending order is handled at the output stage". I think this is where the bug is.

Let's say we have 10k items and the items, after qsort, are arranged in ascending order by some column `[0, 1, 2, ..., 9999]`. Let's also say we only want 20, so this statement becomes `rows = 10000 < 20 ? 10000 : 20` resulting in `rows = 20`. Then the loop on next line will print items `[19, 18, ... 0]` when in reality we wanted it to print items `[9999, 9998, ..., 9980]`.

Guess we were seeing different output due to our test hosts having different # of processes doing fileio, if `rows < output_rows` this will work correctly.

Can do something like:

```
rows_to_print = rows < output_rows ? rows : output_rows;
for (i = rows - 1; i >= (rows - rows_to_print); i--) {
   ...
}
```

Reply:
:facepalm:  You are right. Thanks.

Update the sort_column function instead.

Reply:
confirmed that this fixes the issue